### DISCUSSION TITLE
Raspberry Pi Alternatives

### DISCUSSION THREAD

**Discussion by Smiling-Wanderer** [geezacoleman/OpenWeedLocator#70] [Ideas]
**Summary**
Oh the supply chain woes! With the global lack of Raspberry Pi's, I wanted to explore some alternatives to the Pi.
Over the past couple of weeks, I've tested the LibreComputer Le Potato (AML-S905X-CC) as a possible alternative to the Pi. 
The Potato was able to run the OWL software after some minor changes described in the _Software Setup_ details. 

The Potato is comparable from a budget (35 USD) and performance perspective, but **does not have onboard wifi**, **nor a CSI camera interface**.
Thanks to @geezacoleman, who also figured out that the Potato **does not support USB type 3.0**, so green-on-green detection speed optimization through the Google Coral is a not possible. 
The Raspberry Pi Camera can be adapted to a UVC camera using an Arducam adapter board (IMX477) (40 USD) and then connected via USB to the Potato. 

Though the Potato does seem like a feasible alternative, it may be wiser to move to a more powerful LibreComputer that gives you a little more bang for your buck. The Renegade (ROC-RK3328-CC) could be another feasible alternative, with a cap of 4GB DDR4 RAM for only 40 USD.  The Renegade also has the USB 3.0 that would allow for green-on-green.

There are also other SBC alternatives that are more expensive but would offer significant performance gains. NVIDIA and ASUS both make SBCs that are comparable to the RP 4 that could also definitely be worth checking out. LibreComputer's price points do make their SBCs appealing though...

**Disclaimer**
I am based in the US, so the hardware may or may not be available where you are. 

**Photos**
Le Potato with the adapted RP HQ camera connected via USB
![Le Potato with the adapted RP HQ camera connected via USB](https://user-images.githubusercontent.com/7215005/228692071-1ef49226-4fc7-4d66-bc45-4495d36d090f.jpg)
Close up of the adapted camera setup. The boards are connected by the orange cable that plugs into CSI sockets on the board.
![Close up of the adapted camera setup. The boards are connected by the orange cable that plugs into CSI sockets on the board.](https://user-images.githubusercontent.com/7215005/228692121-7cf44e6f-e6e3-4aa7-8eb5-74269dea0453.jpg)

**Details**
_Additional Hardware_
1. Arducam adapter board (IMX477) for adapting PI HQ CSI to USB. (40 USD)
2. GenBasic WiFi 4 USB Nano Wireless Network Dongle Adapter for Linux (10 USD).

_Software Setup_
LibreComputer has a few distros that are supported on their system. I set up the Potato with 
_Raspbian 11 Bullseye for Libre Computer Boards_. You can find this in the [Potato Downloads](https://libre.computer/products/aml-s905x-cc/) to try and get as close to a RP build as possible for proof-of-concept purposes (and I didn't want to have to adapt to a different distro yet 😃 ). 
Following @geezacoleman's detailed software installation guide worked beautifully. 
If _pycoral_ fails to install during the pip installation, I had success installing the latest version of pycoral with
`pip install pycoral ` 
To use the UVC camera in software, simply access the camera using the video stream and not the picamera module.

_Adapting the Camera_
The lack of a CSI camera interface is quite annoying. Though the camera may be adapted to UVC using the _IMX477_ from Arducam, the overall performance does take a hit and the camera gets bulkier. Follow the instructions provided by Arducam to connect the adapter board to the HQ camera, and then to the Potato. I was able to run the camera at a comfortable 30 FPS without doing much of anything.

**Next Dev Steps**
1. While it was nice to use Raspian for familiarity and compatibility, pivoting to a distro that does a better job of optimizing the LibreComputer's resources seems like the correct path. The _[Ubuntu 22.04.1 Jammy LTS for librecomputer boards](https://hub.libre.computer/t/ubuntu-22-04-1-jammy-lts-for-libre-computer-boards/20)_ might come with some performance gains. 
2. This quick proof-of-concept was really only software and camera. Integration with the rest of the system should be tested thoroughly and performance should be evaluated.
3. The bulkiness of the camera with the adapter board likely will mean a redesign of a portion of the enclosure is needed. 
4. Once we have a functioning software stack, an image should likely be created for ease of deployment.

**Comment by geezacoleman** [comment]
This is fantastic @Smiling-Wanderer ! Really neat write up of the whole process. Any photos of what it looks like with the adapters etc. connected?

 I noticed the Renegade doesn't have the CSI camera interface - is the adapter much of a hassle to work with?
 
 For the green-on-green I also was having a few issues with `pycoral` as well, so ended up making a bash script to try and automate it somewhat. What issues were you having with `pycoral`? It would be nice to have a fairly robust single line command to install/link everything required.

**Reply by Smiling-Wanderer** [reply]
@geezacoleman, great suggestion. I'll edit the original post with some images of what it looks like all put together. 

The adapter is not much of a hassle, it comes with some plastic components so that you can mount it directly behind the RP HQ camera board. See the edited post for a photo. You do have to use a 20 pin ribbon cable to connect the HQ board to the connector, but this is not that much of a hassle. 

The `pycoral` problems were just stemming from pip being unable to find the version that the owl's `requirements.txt` has listed. I believe this is pycoral~=2.0.0. What was a little odd was that pip found it ok on the RP, but not on the potato. It could be that I did something funky in the install process, but there were no other problems for any of the other python packages. It was solved by just allowing pip to find the newest version through `pip install pycoral`. Please note that I'm really not sure whether or not pycoral was utilized beyond import in my tests.

**Comment by david-j-webster** [comment]
is there a step by step for the Jetson Nano?

**Reply by geezacoleman** [reply]
Unfortunately nothing yet  - it's an opportunity we're aware of and would like to explore, but have prioritised the Raspberry Pi currently.
